kernel/userspace.c

```diff
@@ -94,6 +96,144 @@ struct perm_ctx {
 	struct k_thread *parent;
 };
 
+#ifdef CONFIG_DYNAMIC_OBJECTS
+#include <tommyhashlin.h>
+
+extern struct _k_object *_k_object_gperf_find(void *obj);
+extern void _k_object_gperf_wordlist_foreach(_wordlist_cb_func_t func,
+					     void *context);
+
+tommy_hashlin dyn_objects;
+int dyn_objects_init;
+
+static int init_dyn_objects(struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	tommy_hashlin_init(&dyn_objects);
+	dyn_objects_init = 1;
+	return 0;
+}
+
+SYS_INIT(init_dyn_objects, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_OBJECTS);
+
+struct dyn_obj {
+	struct _k_object kobj;
+	tommy_node node;
+	u8_t data[];
+};
+
+static size_t obj_size_get(enum k_objects otype)
+{
+	switch (otype) {
+	case K_OBJ_ALERT:
+		return sizeof(struct k_alert);
+	case K_OBJ_MSGQ:
+		return sizeof(struct k_msgq);
+	case K_OBJ_MUTEX:
+		return sizeof(struct k_mutex);
+	case K_OBJ_PIPE:
+		return sizeof(struct k_pipe);
+	case K_OBJ_SEM:
+		return sizeof(struct k_sem);
+	case K_OBJ_STACK:
+		return sizeof(struct k_stack);
+	case K_OBJ_THREAD:
+		return sizeof(struct k_thread);
+	case K_OBJ_TIMER:
+		return sizeof(struct k_timer);
+	default:
+		return sizeof(struct device);
+	}
+}
+
+static int dyn_obj_cmp(const void *arg, const void *obj)
+{
+	struct _k_object *ko = (struct _k_object *)obj;
+
+	return arg != (void *)(ko->name);
+}
+
+struct _k_object *_k_object_find(void *obj)
```

Comment:
If the gperf lookup fails, the obj parameter has to be validated (otherwise you can dereference any pointer while dyn_obj_cmp() is called).

Reply:
I think what I'm doing is correct here but I renamed the parameters to dyn_obj_cmp() to make it more clear what is and is not being dereferenced.

Reply:
If I understood this right, if you call `_k_object_find(0xdeadbeef)`, the call `dyn_obj_cmp(..., 0xdeadbeef)` will eventually be performed.

Reply:
Indeed.  Looked at how `tommy_hashlin_search()` works, and `dyn_obj_cmp()` will only be called with `obj` pointing to a `struct _k_object` that's in the table.